core/inlooptool.py

```diff
@@ -966,12 +1265,165 @@ def add_build_year_to_surface(self, file_path, field_name="bouwjaar"):
+        output_gpkg = r"C:\Users\ruben.vanderzaag\Documents\Github\bgt-inlooptool\QGIS_plugin\bgtinlooptool\style\output_bgtinlooptool.gpkg"
+        self.copy_and_rename_file(template_gpkg, output_gpkg)
+        
+        print("Saving Pipes layer in gpkg")
+        db_layer = PIPES_TABLE_NAME #"pipes"
+        gpkg_layer = "3. GWSW leidingen" 
+        self._write_to_disk(file_path,db_layer, gpkg_layer)
+
+        print("Saving BGT_inlooptabel layer in gpkg")
+        db_layer =  "bgt_inlooptabel" #RESULT_TABLE_NAME
+        gpkg_layer = "4. BGT inlooptabel"
+        self._write_to_disk(file_path,db_layer, gpkg_layer)
+        self.track_changes(file_path)
+        
+        print("Saving BGT_oppervlak layer in gpkg")
+        db_layer = SURFACES_TABLE_NAME #"bgt_oppervlak"
+        gpkg_layer = "5. BGT oppervlakken"
+        self._write_to_disk(file_path,db_layer, gpkg_layer)
+
+        print("Saving statistics layer in gpkg")
+        db_layer = STATISTICS_TABLE_NAME 
+        gpkg_layer = "6. Statistieken"
+        self._write_to_disk(file_path,db_layer, gpkg_layer)
+        
+        print("Saving calculation settings in gpkg")
+        db_layer = SETTINGS_TABLE_NAME 
+        gpkg_layer = "7. Rekeninstellingen"
+        self._write_to_disk(file_path,db_layer, gpkg_layer)
+
+    
+    def copy_and_rename_file(self,original_file_path, new_file_path):
+        """
+        Copies a file and renames the copy using only sys and os modules.
+
+        :param original_file_path: Path to the original file.
+        :param new_file_path: Path where the new file will be saved.
+        """
+        try:
+            # Read the contents of the original file
+            with open(original_file_path, 'rb') as original_file:
+                content = original_file.read()
+
+            # Write the contents to the new file
+            with open(new_file_path, 'wb') as new_file:
+                new_file.write(content)
+
+            print(f"File copied from {original_file_path} to {new_file_path}")
+        except FileNotFoundError:
+            print(f"The file {original_file_path} does not exist.")
+        except PermissionError:
+            print(f"Permission denied. Unable to copy {original_file_path} to {new_file_path}.")
+        except Exception as e:
+            print(f"An error occurred: {e}")
+            
+    def _write_to_disk(self, file_path, db_layer_name, dst_layer_name): 
+        """Copy self.mem_database to file_path"""
+        # Get the source layer from the memory database
+        self.db_layer = self.mem_database.GetLayerByName(db_layer_name)
+        if self.db_layer is None:
+            raise ValueError(f"Layer '{db_layer_name}' not found in memory database.")
+        
+        # Open the destination GeoPackage in write mode
+        self.dst_gpkg = GPKG_DRIVER.Open(file_path, 1)  # 1 means writable
```

Comment:
Het lijkt me beter om dit met een context manager te doen. Er staat nu wel een dst_gpkg = None aan het eind, maar nu is er wel gevaar dat het bestand nog open is als er iets in de code fout gaat.

context manager ziet er vast bekend uit:

with GPKG_DRIVER.Open(file_path, 1) as self.dst_gpkg:
..hier de code

Reply:
Gedaan. Werkt goed!